File: 2022/day25/main.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <vector>
#include <math.h>
#include <deque>

#define LOG(x) std::cout << x << std::endl;
using ll = long long;

std::string convertToSNAFU(ll val)
{
    std::string str = std::to_string(val);
    std::deque<char> q;
    ll tmp = val;
    std::string chars("=-012");
    while (tmp != 0)
    {
        ll div = floor((tmp + 2) / 5);
        int idx = (tmp + 2) % 5;
        tmp = div;
        q.push_front(chars.at(idx));
    }
    std::string res;
    while (q.size() > 0)
    {
        char c = q.front();
        res.push_back(c);
        q.pop_front();
    }
    return res;
}

ll convertToInt(std::string snafu)
{
    ll res = 0;
    for (int i = snafu.size() - 1; i >= 0; i--)
    {
        char c = snafu.at(i);
        int pos = snafu.size() - i - 1;
        int val = c - '0';
        if (c == '-')
        {
            val = -1;
        }
        else if (c == '=')
        {
            val = -2;
        }
        ll cal = val * pow(5, pos);
        res += cal;
    }
    return res;
}

int main()
{
    std::cout << "Hello world" << std::endl;
    std::string line;
    // std::ifstream file("small");
    std::ifstream file("input");

    std::vector<int> nums;
    ll toBeConv = 0;
    while (getline(file, line))
    {
        std::cout << line << " -> ";
        ll num = convertToInt(line);
        nums.push_back(num);
        std::cout << num << std::endl;
        toBeConv += num;
    }
    LOG("To be converted to SNAFU: " << toBeConv)
    std::string conv = convertToSNAFU(toBeConv);
    LOG("Part 1: " << conv)
}

File: 2022/day25/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func convertToDecimal(snafu string, ch chan int) {
	for i := len(snafu) - 1; i >= 0; i-- {
		var val int
		if snafu[i] == '-' {
			val = -1
		} else if snafu[i] == '=' {
			val = -2
		} else {
			val = int(snafu[i] - '0')
		}
		ch <- val * int(math.Pow(5, float64(len(snafu)-i-1)))
	}
}

func main() {
	f, err := os.Open("input")
	defer f.Close()
	check(err)
	scanner := bufio.NewScanner(f)
	ch := make(chan int)
	nExpect := 0
	for scanner.Scan() {
		line := scanner.Text()
		nExpect += len(line)
		go convertToDecimal(line, ch)
	}

	toBeConv := 0
	for nExpect > 0 {
		toBeConv += <-ch
		nExpect--
	}

	fmt.Println("To be conv:", toBeConv)

	var chars [5]rune = [5]rune{'=', '-', '0', '1', '2'}
	arr := make([]rune, 20)
	for toBeConv != 0 {
		pf := toBeConv + 2
		toBeConv = pf / 5
		arr = append(arr, chars[pf%5])

	}
	rev := make([]rune, len(arr)+1)
	for i := len(arr) - 1; i >= 0; i-- {
		rev[len(arr)-i] = arr[i]
	}
	res := string(rev)
	fmt.Println("Part 1:", res)
}
